### PR TITLE
Update fused kernels and call _safe_softmax from SDPA

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133373
* __->__ #131863

# Summary
Changes the stance of SDPA on what to do for fully masked out rows

## Current Behavior
Several PyTorch users have expressed frustration over this issue:
- https://github.com/pytorch/pytorch/issues/41508
- https://github.com/pytorch/pytorch/issues/103749
- https://github.com/pytorch/pytorch/issues/103963

These are significant issues with extensive discussion but no satisfactory resolution. The PyTorch team's consensus, as stated here:
https://github.com/pytorch/pytorch/issues/24816#issuecomment-524415617

Can be paraphrased as follows:

When passing in fully masked out rows, attention becomes ambiguous. We have two main options:

1. Uniformly attend to all values:
   ```python
   scores[masked_out_rows] = 1 / len(row)
   out[masked_out_rows] = 1 / len(row) * value
   ```

2. Decide that attention between no queries (masked) and no keys (masked) is meaningless:
   ```python
   output[fully_masked_rows] = NaN
   ```

We went with option 2. Partially because it was easier to implement, but also people argued that users can slice the output to remove the NaNs:
``` Python
>fill_value = -float("inf") 
>row0 = torch.randn(4)
>row1 = torch.tensor([(fill_value for _ in range(4)])
>matrix = torch.stack([row0, row1]).requires_grad_(True)
>out = torch.softmax(matrix, 1)
>out = out[0]
>print(out)
tensor([0.5377, 0.2729, 0.0692, 0.1201])
```
Cool, problem solved. But what happends when you call backwards..
```Python
>out.backward(torch.ones_like(out))
>print(matrix.grad)
tensor([[3.0957e-08, 1.4157e-08, 7.7802e-10, 1.3713e-08],
        [       nan,        nan,        nan,        nan]])
```
Those pesky NaNs are back!

## Why do we see NaNs today?

The core of the problem revolves around using softmax function in sdpa:

```python
> row = torch.tensor([(-float("inf")) for _ in range(4)])
> torch.softmax(row, 0)
tensor([nan, nan, nan, nan])
```

## Quick Aside: Masking in Attention

Attention itself doesn't have a concept of masking. The `sdpa` function has an argument called `attn_mask`, which would be more accurately named `attn_bias`. This is because we don't actually "mask" entries when computing attention. Instead, due to implementation details([performance](https://github.com/pytorch/pytorch/issues/25110#issuecomment-524519087)), we add a value to the masked-out query/key pairs.

We use a large negative number (typically -inf) to decrease the attention weight, as softmax assigns more weight to larger values.

## Alternative Approaches

If we use a very large negative number instead of -inf:

```python
> row = torch.tensor([(-1e6) for _ in range(4)])
> torch.softmax(row, 0)
tensor([0.2500, 0.2500, 0.2500, 0.2500])
```
However if users always remembered to "slice" out their outputs i.e.:
```Python
>fill_value = -1e6
>...
>out.backward(torch.ones_like(out))
>print(matrix.grad)
tensor([[-0.0563, -0.0564,  0.1613, -0.0486],
        [ 0.0000,  0.0000,  0.0000,  0.0000]])
```
This would bring us back into a better state.

## A Third Option

We don't necessarily need to alter the behavior of softmax for -inf or very large negative numbers. The fundamental goal is to exclude certain query/key pairs from attention, regardless of the underlying implementation.

This PR implements the new semantic for masking w/ attention in fully masked-out rows:
```python
out[masked_out_rows] = 0
```

**Important Note**: This idea isn't entirely new. The [MaskedTensor](https://pytorch.org/tutorials/prototype/maskedtensor_overview#safe-softmax) prototype, a tensor subclass, was designed to handle such cases. However, it remains a prototype feature and hasn't gained widespread adoption.

## Details
This PR stack does 3 things:
1. Adds a PRIVATE _safe_softmax op
2. Updates semantic for flash_cpu fused kernel
3. Updates semantic for efficient_cuda fused kernel

_safe_softmax is not supposed to be used generically and is only meant to be used within the context of SDPA. Due to this fact instead of decomposing softmax and checking for -inf rows we instead "cheat" and use nan_to_num.

Why I think this is okay? (please find a counter point if avail)
There are multiple ways NaNs can emerge. For the fully masked out rows case nan_to_num works. But what if there were other NaNs, wouldn't this silently remove them? 

The only case that this can happen is if the input itself had a NaN or an Inf
For example: 
```Python
a = torch.ones([4], requires_grad=False, dtype=torch.float16)
a[1] = torch.finfo(torch.float16).max
print(a.softmax(-1))
```
Will return 
`tensor([0., 1., 0., 0.], dtype=torch.float16)`

Where 
```Python
a = torch.ones([4], requires_grad=False, dtype=torch.float16)
a[1] = float("inf")
a.softmax(-1)
```
returns:
`tensor([nan, nan, nan, nan], dtype=torch.float16)`

If we dont want to even allow for the possibility of "inf" or "NaN" attention scores to be converted to 0 then we can implemented it something like this

```Python
max = torch.max(a, dim=-1, keepdim=True)
exp = torch.exp(a - max.values)
denom = torch.sum(exp, dim=-1, keepdim=True)
softmax = exp / denom
softmax = torch.where(max.values == float('-inf'), 0.0, softmax)
```
however we would be paying for this in math performance.

## Why Now
I think one point that has substantially changed where PyTorch should lie on this argument is the fact that we have fused implementations for SDPA now. And these fused implementations allow us to easily and performantly support this new semantic. 

cc @ezyang @gchanan @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @Valentine233

Differential Revision: [D61418679](https://our.internmc.facebook.com/intern/diff/D61418679)

### CODE REVIEW

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 1]
```diff
@@ -422,6 +422,12 @@ void cpu_flash_attention(
...
       }
+      // We set fully masked out rows to 0, for this masked out rows the sum is 0, so we need to set it to 1
+      for (const auto row : c10::irange(qBlockSize)){
+        qk_max_data[row] = qk_max_data[row] == -std::numeric_limits<accum_t>::infinity() ? 0 : qk_max_data[row];
+        qk_sum_data[row] = qk_sum_data[row] == 0 ? 1 : qk_sum_data[row];
+      }
+
       // dst <- dst / sum[row]
       // reorder MHA output with strides
       for (int64_t row = 0; row < qBlockSize; ++row) {
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:433]: nit: shall we fuse these two loops?

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 2]
```diff
@@ -422,9 +422,15 @@ void cpu_flash_attention(
...
       }
+
       // dst <- dst / sum[row]
       // reorder MHA output with strides
       for (int64_t row = 0; row < qBlockSize; ++row) {
+        // Row sums for full masked out rows are 0, we set them to 1
+        // in order to avoid NaNs in the output and instead set fully
+        // masked out rows to 0
+        qk_max_data[row] = qk_max_data[row] == -std::numeric_limits<accum_t>::infinity() ? 0 : qk_max_data[row];
+        qk_sum_data[row] = qk_sum_data[row] == 0 ? 1 : qk_sum_data[row];
```
Comment (jbschlosser) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:433]: dumb Q: why does setting the sums to 1 result in the output having 0s for fully masked-out rows?

**Review by jbschlosser — COMMENTED** [review]
Nice! do you need any changes to flash backward as well to fix the gradients?

**Review by jbschlosser — APPROVED** [review]

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 2 above)
Comment (drisspg) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:433]: qk_sum_data is the the `sumexp(i-max)` When we set this. 

So  `qk_max_data[row] =  ..` This makes converts the output from nan to 0
and then 
`qk_sum_data[row]` This makes so that instead dividing by 0 on line 434 for normalizing we instead divide by 1

**Comment by drisspg** [comment]
The test failures seems to be coming from:
https://github.com/pytorch/xla/blob/1ed2626725c47fcf0979e0790a219bf3b9de4ebd/test/stablehlo/test_composite.py#L95

**Comment by drisspg** [comment]
closing for this stack:
https://github.com/pytorch/pytorch/pull/131863

**Comment by drisspg** [comment]
The test failures seems to be coming from:
[pytorch/xla@1ed2626/test/stablehlo/test_composite.py#L95](https://github.com/pytorch/xla/blob/1ed2626725c47fcf0979e0790a219bf3b9de4ebd/test/stablehlo/test_composite.py#L95)

**Comment by albanD** [comment]
@pytorchbot revert -m "Broke forward AD tests in main" -c nosignal

See https://hud.pytorch.org/pytorch/pytorch/commit/e61def65d5c6268e79f52776f75277ee60f01462 as an example.

**Review by Chillee — APPROVED** [review]

**Comment by drisspg** [comment]
@drisspg has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61418679).
